### PR TITLE
Implement external classes

### LINKED ISSUE #250: How to create a nestable R7 class?
Sometimes it's useful to be able to nest objects of the same type inside each other.  Is there a recommended way for doing this with R7?

##  Naive approach
It's good that this errors; if you ever tried to instantiate this you'd end up in an infinitely recursing loop.

``` r
nested <- new_class("nested",
  properties = list(child = nested)
)
#> Error in as_properties(properties): object 'nested' not found
```

<sup>Created on 2022-08-14 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

## `new_union()`
It would be nice if this worked, allowing you to specify at least one other type that can be a "base case". Preferably while still protecting you from infinite recursion by not permitting the nested class to be the default.

``` r
nested <- new_class("nested",
  properties = list( child = new_union(NULL, nested) )
)
#> Error in lapply(x, as_class): object 'nested' not found
```

<sup>Created on 2022-08-14 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

## Define it twice

This doesn't error and lets me nest objects, but `properties` holds on to the original class definition (`inner`'s child has no `name` or `child` properties, since the original definition of `nested` had no properties).  I don't know what problems that might cause (and it prevents the infinite recursion problem), but I don't like the feel of it.

``` r
nested <- new_class("nested")

nested <- new_class("nested",
  properties = list( 
    name = class_character,
    child = nested )
)

inner <- nested()
inner
#> <nested>
#>  @ name : chr(0) 
#>  @ child: <nested>

outer <- nested(name = "outer", child = inner)
outer
#> <nested>
#>  @ name : chr "outer"
#>  @ child: <nested>
#>  .. @ name : chr(0) 
#>  .. @ child: <nested>
```

<sup>Created on 2022-08-14 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

# Use a parent class

This seems to work and I haven't thougth of anything that would go wrong.  It's mildly annoying to have the extra parent class cluttering things up if I don't have any other reason for it to exist, but I can live with it :)

``` r
nested_parent <- new_class("nested_parent")

nested <- new_class("nested", parent = nested_parent,
  properties = list( 
    name = class_character,
    child = nested_parent )
)

inner <- nested()
inner
#> <nested>
#>  @ name : chr(0) 
#>  @ child: <nested_parent>

outer <- nested(name = "outer", child = inner)
outer
#> <nested>
#>  @ name : chr "outer"
#>  @ child: <nested>
#>  .. @ name : chr(0) 
#>  .. @ child: <nested_parent>
```

<sup>Created on 2022-08-14 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

So, 2 requests:
1. It'd be nice to have documentation on the recommended way to create nestable classes. Is using a parent class the way to go?
2. Any chance some behind-the-scenes wizardry could make something like the `new_union` syntax work?

### LINKED ISSUE #573: Defining generic methods for suggested packages
Hello everyone,

When building a package that is exporting a generic function, what is the proper way to export a method for a class within a "Suggests" package?

For example, if package `foo` exports the `foo()` generic, but lists package `bar` as a Suggests, what would be the proper way to register methods for classes within `bar`?

Would it be something to the effect of:

```r
# from the foo package
.onLoad <- function(lib, pkg) {

  if (isNamespaceLoaded("bar")) {
    method(foo, bar::class_bar) <- function(x) {cat("we dispatched on `bar::class_bar`")}
  } else {
    setHook(packageEvent("bar", "onLoad"), function(){
      method(foo, bar::class_bar) <- function(x) {cat("we dispatched on `bar::class_bar`")}
    })
  }
}

```

### PR DESCRIPTION
**Pull request by hadley**
I think the use of external is a little bit misleading here, because these are really "delayed" classes. But external matches external generic, which exist for a similar reason.

Fixes #250. Fixes #573.

Note this does not solve #317, since currently creating a subclass will force loading of the package via `class_constructor()`. But I think can fix that after #694 by using `...` to call constructors from other packages. That will decouple the superclass arguments from the subclass in a way that won't cause `R CMD check` to fail with a codoc mismatch if the parent args change.

### CODE REVIEW

**Comment by hadley** [comment]
@t-kalinowski @lawremi I think this is now ready for your feedback.

**Review by lawremi — APPROVED** [review]
This is a very useful feature. Implementation looks good.

**Review by t-kalinowski — APPROVED** [review]
I pushed tests and fixes for some edge cases. Can you please take a look and see if the approach makes sense?

An open question is what behavior we want for property override checks when an external class is from a package that is not loaded yet. The current fix avoids treating an unresolved external class as matching every child type, but it's possible we want to take a different approach here (e.g., defer until the package is loaded, or load the package eagerly)

File: R/external-class.R [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+find_external_class <- function(x) {
+  ns <- asNamespace(x$package)
+  if (exists(x$name, envir = ns, inherits = FALSE)) {
+    obj <- get(x$name, envir = ns, inherits = FALSE)
+    if (is_external_class_match(obj, x)) {
+      return(obj)
+    }
+  }
+
+  # Also consider cases where the constructor isn't named the same as the class
```
Comment (hadley) [inline R/external-class.R]: Are we sure that this is a case that we want to support? Why do we want to encourage people to have a mismatch between constructor and class name?

File: R/external-class.R [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+#'
+#'   ```R
+#'   tree_stub <- new_external_class("mypkg", "tree")
+#'   new_class("tree", properties = list(child = NULL | tree_stub))
+#'   ```
+#'
+#' Make sure to call [S7_on_load()] in your package's `.onLoad()` so that
+#' deferred method registrations fire when the relevant package is loaded.
+#'
+#' External classes can not currently be used as parents in [new_class()].
```
Comment (hadley) [inline R/external-class.R:31]: Made this restriction more clear. Next step is to resolve #609, and #317.

**Review by t-kalinowski — COMMENTED** [review]
I opened a PR with some proposed changes to this branch: https://github.com/RConsortium/S7/pull/710

Most of the changes address various edge cases raised by automated review.

One change I want to call out is that I'm proposing that external S7 classes must be exported by a package using the same symbol name as the class's `@name`. This makes resolution much simpler and faster. It also makes the code easier to reason about, both internally in S7 and in user code, and avoids introducing a bunch of complexity. In particular, we don't need to go on a fishing expedition looking for an exact match, which could potentially force a bunch of lazy bindings.

I also tightened up a few edge cases around this:

- external class resolution now has separate errors for missing packages, too-old package versions, and binding mismatches
- external classes work in union arms, including self-referential and mutually recursive class definitions
- child properties can narrow an external parent class to a loaded subclass
- subclassing against an unloaded external class now fails at the point where we need to resolve it
- method introspection and `method_explain()` now use the same required-resolution behavior
- unload cleanup handles the case where the external class package has already unloaded before the downstream package unloads

I also changed one of the tests from using `local_mocked_bindings()` to `local_package()`, to avoid needing to define `getNamespaceVersion` in the package namespace just to mock it during tests -- that seemed to me like test scaffolding leaking into production code.
